File: 백준/Gold/1987. 알파벳/알파벳.java
import java.io.*;
import java.util.*;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
	static StringTokenizer st;
	static int R,C,ans;
	static char[][] arr;
	static boolean[] vis;
	static final int[] dr = {-1,0,1,0};
	static final int[] dc = {0,1,0,-1};
	
	public static void main(String[] args) throws IOException {
		st = new StringTokenizer(br.readLine());
		R = Integer.parseInt(st.nextToken());
		C = Integer.parseInt(st.nextToken());
		arr = new char[R][C];
		vis = new boolean[26];
		for(int r=0; r<R; r++) {
			String line = br.readLine();
			for(int c=0; c<C; c++) arr[r][c] = line.charAt(c);
		}
		vis[arr[0][0]-'A']=true;
		dfs(0,0,1);
		bw.write(ans+"");
		bw.close();
		br.close();
	}
	
	static void dfs(int r, int c, int cnt) {
		ans = Math.max(ans, cnt);
		for(int dir=0; dir<4; dir++) {
			int nr = r+dr[dir];
			int nc = c+dc[dir];
			if(nr<0||nr>=R||nc<0||nc>=C) continue;
			int idx = arr[nr][nc]-'A';
			if(!vis[idx]) {
				vis[idx]=true;
				dfs(nr,nc,cnt+1);
				vis[idx]=false;
			}
		}
	}
}

File: 백준/Gold/1987. 알파벳/알파벳.cc
#include <bits/stdc++.h>
using namespace std;
int R,C,ret=1;
string S;
char a[21][21];
bool vis[26];
const int dr[] = {-1,0,1,0};
const int dc[] = {0,1,0,-1};
void dfs(int r, int c, int cnt) {
    ret=max(ret,cnt);
    for(int dir=0; dir<4; dir++) {
        int nr=r+dr[dir];
        int nc=c+dc[dir];
        if(nr<0||nr>=R||nc<0||nc>=C) continue;
        if(!vis[a[nr][nc]-'A']) {
            vis[a[nr][nc]-'A']=true;
            dfs(nr,nc,cnt+1);
            vis[a[nr][nc]-'A']=false;
        }
    }
}
int main() {
    ios_base::sync_with_stdio(0); cin.tie(0);
    cin >> R >> C;
    for(int r=0; r<R; r++) {
        cin >> S;
        for(int c=0; c<C; c++) a[r][c]=S[c];
    }
    vis[a[0][0]-'A']=true;
    dfs(0,0,1);
    cout << ret << '\n';
    return 0;
}